### PR TITLE
Group measures by resource metadata history

### PR DESCRIPTION
**Pull request by pedro-martins**
Problem description
===============

When we are aggregating measures using Gnocchi, we can group the measures by resource's metadata. However, Gnocchi does not group the measures taking into account the resource's metadata history; so if some resource's metadata changed recently, it will be grouped using only the last metadata value. 

Imagine the following situation:

I have a metric with metadata called "name", so I take a resource that is monitored by this metric and I change the resource's name from "test" to "test1" at 14:45:00. Then I call the Gnocchi's Aggregation API with an interval from 11 to 17 hours grouping by "name" and it returns this:

| group                  | name       | timestamp                    | granularity | value               |
|------------------------|------------|------------------------------|-------------|---------------------|
| name: test1 | aggregated | 2020\-03\-26T11:00:00\+00:00 | 3600\.0     | 101 |
| name: test1 | aggregated | 2020\-03\-26T12:00:00\+00:00 | 3600\.0     | 104 |
| name: test1 | aggregated | 2020\-03\-26T13:00:00\+00:00 | 3600\.0     | 98  |
| name: test1 | aggregated | 2020\-03\-26T14:00:00\+00:00 | 3600\.0     | 100  |
| name: test1 | aggregated | 2020\-03\-26T15:00:00\+00:00 | 3600\.0     | 97  |
| name: test1 | aggregated | 2020\-03\-26T16:00:00\+00:00 | 3600\.0     | 109 |

Here we have a problem. For the system consuming Gnocchi API, it will look like the resource has the same attributes in all of the timestamps, when that is not true. This can have a negative impact on billing systems that rely on Gnocchi as the time series database (e.g. CloudKitty).

Proposal
=======

I propose to allow users to specify if they want or not to group their measurements taking into account the resource's metadata history; if they want, they can just add a "use_history=true" parameter in their API request and Gnocchi will do it for them. If users do not specify the "use_history" in the API, the following property will be used instead.

```property
[api]
group_measures_using_history = True/False
```

If the property or the query parameter are not defined, the "use_history" value assumes False (for backward compatibility). The new response expected with this proposal using the same situation listed in `Problem description` will be (using group_measures_using_history=true):

| group                  | name       | timestamp                    | granularity | value               |
|------------------------|------------|------------------------------|-------------|---------------------|
| name: test  | aggregated | 2020\-03\-26T11:00:00\+00:00 | 3600\.0     | 101 |
| name: test  | aggregated | 2020\-03\-26T12:00:00\+00:00 | 3600\.0     | 104 |
| name: test  | aggregated | 2020\-03\-26T13:00:00\+00:00 | 3600\.0     | 98  |
| name: test  | aggregated | 2020\-03\-26T14:00:00\+00:00 | 3600\.0     | 75  |
| name: test1 | aggregated | 2020\-03\-26T14:00:00\+00:00 | 3600\.0     | 25  |
| name: test1 | aggregated | 2020\-03\-26T15:00:00\+00:00 | 3600\.0     | 97  |
| name: test1 | aggregated | 2020\-03\-26T16:00:00\+00:00 | 3600\.0     | 109 |

As you can see, there are 2 entries at 14:00; it happens because the resource's name was changed at 14:45, that is equivalent to 75% of the granularity window (that is 1 hour), so the measure named "test" will take 75% of the total measurement of the granularity window, and the measure named "test1" will take the remaining 25% of the measurement value.

### CODE REVIEW

**Comment by jd** [comment]
Hi and thank you guys!
Right now the number one issue is https://github.com/gnocchixyz/gnocchi/issues/1049 as you might have seen.

I don't see the CI running on this PR so I feel like it really does not work anymore or has been disabled?

I feel like there's really some work to be done here before we can merge anything.

This PR seems quite large so it might require some time to review, especially since nobody's usually around.

**Comment by tobias-urdin** [comment]
quite a large change, can you please recheck/reopen to get a new CI check and we can try getting this back on the road.

**Comment by chungg** [comment]
@mergify rebase

File: gnocchi/indexer/sqlalchemy_base.py
Comment (jd) [inline gnocchi/indexer/sqlalchemy_base.py:202]:
```suggestion
        return str(self.jsonify())
```

File: gnocchi/opts.py
Comment (jd) [inline gnocchi/opts.py:196]: I find it always weird to control API behaviour via global config options.
Why isn't that a only flag passed to the API?

File: gnocchi/rest/aggregates/api.py
Comment (jd) [inline gnocchi/rest/aggregates/api.py:190]: Yeah I don't think a global option should change a default behaviour.

File: run-upgrade-tests.sh
Comment (jd) [inline run-upgrade-tests.sh:73]: unrelated :)

**Review by jd — CHANGES_REQUESTED** [review]
This is pretty long and complex, so I'm unable to review it deeply. But a quick glance make it looks good so I think we can merge.
The only thing I'd change is the option which I would remove.

File: gnocchi/opts.py
Comment (pedro-martins) [reply gnocchi/opts.py:196]: This was done to avoid changing every system that uses this API to get the new behavior, so, instread of changing lots of systems to use the new API flag, we just set the configuration in gnocchi, and if some system for some reason need the old behavior, it can simple override the config with the API flag.

File: gnocchi/indexer/sqlalchemy_base.py
Comment (pedro-martins) [reply gnocchi/indexer/sqlalchemy_base.py:202]: nice catch, thanks.

File: run-upgrade-tests.sh
Comment (pedro-martins) [reply run-upgrade-tests.sh:73]: yea, it was added for some debugging purpose and I forgot to remove it, thanks.

File: gnocchi/opts.py
Comment (jd) [reply gnocchi/opts.py:196]: This is not a good approach as an API should never ever change and break existing system.

Comment (pedro-martins) [reply gnocchi/opts.py]: I will remove the config

File: gnocchi/rest/aggregates/api.py
Comment (pedro-martins) [reply gnocchi/rest/aggregates/api.py]: Ok, no problem, I will remove it.

**Review by chungg — COMMENTED** [review]
don't understand this enough to approve, but not against since it seems leave current functionality as is.

**Comment by tobias-urdin** [comment]
Possible that something changed with the new pyparsing version just released https://github.com/pyparsing/pyparsing/releases - I don't have the time investigate this straight away though, will check later.

File: gnocchi/rest/aggregates/measures_grouper.py
Comment (pedro-martins) [reply gnocchi/rest/aggregates/measures_grouper.py:264]: This "get_data_map" function will return a map (dict) within all group`s measures grouped by timestamp (date). 

If the group has no measures yet, then it will set the group entry and return the dict reference to be filled with the measures in  the "get_measures" function. 

If the group has no ID them it will return all measures grouped by group and grouped by date

File: gnocchi/rest/aggregates/measures_grouper.py
Comment (pedro-martins) [reply gnocchi/rest/aggregates/measures_grouper.py]: ok

File: gnocchi/rest/aggregates/measures_grouper.py
Comment (pedro-martins) [reply gnocchi/rest/aggregates/measures_grouper.py]: yes

File: gnocchi/rest/aggregates/measures_grouper.py
Comment (pedro-martins) [reply gnocchi/rest/aggregates/measures_grouper.py]: removed

**Comment by tobias-urdin** [comment]
Can this conflict be resolved and then we can make have a another look at this and merge it? Thanks!

**Comment by tobias-urdin** [comment]
pyparsing changed the output so this needs a change like was done in https://github.com/gnocchixyz/gnocchi/pull/1185/commits/70da17a90a13151a307c51bc26bf6994c1ff222a

**Comment by pedro-martins** [comment]
Thanks tobias, fix applied.

File: gnocchi/rest/aggregates/api.py
Comment (tobias-urdin) [inline gnocchi/rest/aggregates/api.py:223]: Please add this back into a try statement and catch the IndexerException and return 400 as was done before. You can move it inside the try statement below.

**Review by tobias-urdin — CHANGES_REQUESTED** [review]
A small thing to change otherwise LGTM, maybe @jd you can take a final look as well?

File: gnocchi/tests/test_measures_grouper.py
Comment (tobias-urdin) [inline gnocchi/tests/test_measures_grouper.py:1]: I can't see that this tests is ever executed in CI testing?

**Review by tobias-urdin — CHANGES_REQUESTED** [review]

File: gnocchi/rest/aggregates/api.py
Comment (pedro-martins) [reply gnocchi/rest/aggregates/api.py]: Yes, sure, I missed that. 
Thanks.

File: gnocchi/tests/test_measures_grouper.py
Comment (pedro-martins) [reply gnocchi/tests/test_measures_grouper.py:1]: I checked, indeed, they were not executed, I fixed it as well.

Thanks again.

File: gnocchi/rest/aggregates/api.py
Comment (tobias-urdin) [inline gnocchi/rest/aggregates/api.py:17]: Please dont import distutils here but use gnocchi.rest.api.get_bool_param as already used in this file.

File: gnocchi/rest/aggregates/api.py
Comment (tobias-urdin) [inline gnocchi/rest/aggregates/api.py:189]: use api.get_bool_param

**Review by tobias-urdin — CHANGES_REQUESTED** [review]
Some more feedback on this patch now that I've managed to fix the CI again. My intent is to merge this PR now after walking through it but have some more things, see the review.

@jd my intent is to merge this, please provide feedback on this patch if you have any.

**Comment by tobias-urdin** [comment]
Please rebase against master as well to get all the changes for the CI to work as well.

File: gnocchi/rest/aggregates/api.py
Comment (pedro-martins) [reply gnocchi/rest/aggregates/api.py:17]: sure, thanks.

File: gnocchi/rest/aggregates/api.py
Comment (pedro-martins) [reply gnocchi/rest/aggregates/api.py]: Ok, thanks :slightly_smiling_face:

File: gnocchi/rest/aggregates/api.py
Comment (pedro-martins) [reply gnocchi/rest/aggregates/api.py]: We dont have a reference to LOG in this scope, I will add one using the 'logging' module, should I use the 'daiquiri' module instead? or is the 'logging' ok?

Also, for this generic exception, catching, in some cases we have already handled exceptions (that are the httpexceptions one) so I will add a new except to ignore these ones to avoid breaking some tests and API compatibilities

Comment (tobias-urdin) [reply gnocchi/rest/aggregates/api.py]: I think logging is OK since it's used in other places. I don't think you need to introduce anything other than what I suggested since it's just a nicer way of doing it and it would catch anything.

Basicly, just throwing the exception would cause the same behavior anyway, the HTTP 500 error. It's just a matter if we should "leak" the exception message in the response or not. Maybe we shouldn't, what do you think?

Comment (tobias-urdin) [reply gnocchi/rest/aggregates/api.py]: Thinking about it, we should probably not return the exception, only log it. And you can skip the weboc.HTTPExceptions

File: gnocchi/rest/aggregates/api.py
Comment (jd) [inline gnocchi/rest/aggregates/api.py:252]: probably not the right place to log

**Review by jd — COMMENTED** [review]
honestly I don't have enough time to review this, I'm sorry

File: gnocchi/rest/aggregates/api.py
Comment (pedro-martins) [reply gnocchi/rest/aggregates/api.py:252]: Hmm, maybe, where do you suggest I should put this log? I cant think in another place. Thanks :slightly_smiling_face:

File: gnocchi/rest/aggregates/measures_grouper.py
Comment (chungg) [inline gnocchi/rest/aggregates/measures_grouper.py:88]: hmm... the performance of this is a bit concerning. it seems to be going through each timeseries point by point and trying to slot in to a group (in python)

**Review by chungg — COMMENTED** [review]
i imagine this would only make sense for the rate/delta metrics? kind of wondering how it'd be used for other metric types? i keep thinking arbitrarily dividing something like cpu utilisation's aggregated measure across two groups doesn't make sense and dividing something like disk.storage.bytes even less.

File: gnocchi/rest/aggregates/api.py
Comment (tobias-urdin) [reply gnocchi/rest/aggregates/api.py:336]: Including the exception however could include information about internals that should not be exposed, so either we should just leave it and give a 500 error and treat it as a bug, or return 400 and hope that it's valid. I'm more for 500 than returning 400 and it's potentially not an input issue.

File: gnocchi/rest/aggregates/api.py
Comment (tobias-urdin) [reply gnocchi/rest/aggregates/api.py:252]: See my comment below on the same point.

File: gnocchi/rest/aggregates/api.py
Comment (pedro-martins) [reply gnocchi/rest/aggregates/api.py:336]: Yes, but in that case, we have some problems to discuss here:
 - About line 252
   - If we catch the Exception and abort it with an error 500 or 400, in some cases the Exception is an instance of `weboc.HTTPExceptions` that is some already handled exception, so in the case this Exception raises an HTTP 400 and we catch it to an HTTP 500, it will be an inconsistency. To try to solve it, we could catch an `weboc.HTTPExceptions` before catching the Exception itself and and ignore it.
   - If we dont catch the generic Exception, we will delegate to who called this method to handle it, and its ok in my opinion, and that is what we are doing here right now, except that we are logging the Exception before throwing it. This is the initial behavior of this procedure (we only added the logging stuff).
 - About line 331. No reason, I missed this one, sorry.

File: gnocchi/rest/aggregates/measures_grouper.py
Comment (pedro-martins) [reply gnocchi/rest/aggregates/measures_grouper.py:26]: Sure, no problem, thanks :slightly_smiling_face:

File: gnocchi/rest/aggregates/api.py
Comment (pedro-martins) [reply gnocchi/rest/aggregates/api.py:33]: Ok, thanks.

File: gnocchi/rest/aggregates/measures_grouper.py
Comment (pedro-martins) [reply gnocchi/rest/aggregates/measures_grouper.py:142]: We created a class just to try to organize things up when  codding this one. Is there any problem? We could change it if necessary.

File: gnocchi/rest/aggregates/measures_grouper.py
Comment (chungg) [reply gnocchi/rest/aggregates/measures_grouper.py:88]: got it. (wasn't aware nova kept same id when flavour changes). i do understand the problem you're trying to solve, i guess my question is what the correct behaviour is to attribute ownership when dividing up an measure.

for instance, if we're capturing cpu.time|storage.bytes which is a gauge metric that (for the most part) only grows. if we split, it with current logic say it might end up being:
- 00:00:00, value 0, meta1
- 10:00:00, value 10, meta1
- 11:00:00, value 1.1; meta1(11:00 - 11:06)
- 11:00:00, value 8,8: meta2(11:06 - 11:54)
- 11:00:00, value 1.1, meta3 (11:54 - 12:00)

which would be an odd value considering it was meta1 for 11hrs and meta2 for 48min but the values are near identical. maybe i'm over-complicating this.

File: gnocchi/rest/aggregates/api.py
Comment (pedro-martins) [reply gnocchi/rest/aggregates/api.py:252]: It will not return an empty result, it will raise the Exception again (and maybe return error 500, or 404, it will depends how the caller is handling exceptions, but it is how it was implemented before (except by the log stuff)

File: gnocchi/rest/aggregates/measures_grouper.py
Comment (pedro-martins) [reply gnocchi/rest/aggregates/measures_grouper.py:88]: Got your point, and yes, your are right, this kind of behavior does not make much sense when using gauge metrics, they are cumulative. 

You should not configure metadata sensitive resources using gauge metrics when you have floating metadata (each possible metadata should be a resource itself to gauge make sense), because any time the metadata changes you should reset the measures values to zero (because it means that this is a new one) and pause the old metadata collecting until the resource changes its metadata again to the old value (togging between metadata values).

For these cases, if you want to use gauge metrics, you should create one metric for any possible metadata value in the resource and measure them separated in your data polling service. 

So if this is the case, you could continue using the normal aggregation API without sending the parameter "user_history" in the request, so the default behavior will me maintained.

**Review by chungg — COMMENTED** [review]
if i'm honest, i'm not entirely sure this is the ideal implementation or behaviour but i'm also not as involved anymore and can see it's a valid use case so am not against it. if others are ok with it, i would suggest docs and some api tests.

**Comment by pedro-martins** [comment]
> if i'm honest, i'm not entirely sure this is the ideal implementation or behaviour but i'm also not as involved anymore and can see it's a valid use case so am not against it. if others are ok with it, i would suggest docs and some api tests.

I added some gabbit tests and rest docs regarding the changes.

**Review by tobias-urdin — APPROVED** [review]

**Pull request merged**
